domain/secretbackend/state/state.go

```diff
@@ -454,122 +456,63 @@ func (s *State) GetModelSecretBackend(ctx context.Context, modelUUID coremodel.U
-    LEFT JOIN user cu ON m.owner_uuid = cu.uuid
-    JOIN cloud_credential cc ON m.cloud_credential_uuid = cc.uuid
-    LEFT JOIN cloud_credential_attributes cc_attr ON cc_attr.cloud_credential_uuid = cc.uuid
-WHERE m.model_uuid = $M.uuid`,
-		sqlair.M{}, CloudRow{},
-		credentialstate.Credential{}, credentialstate.CredentialAttribute{},
-	)
+	stmt, err := s.Prepare(q, sqlair.M{}, modelCloudAndCredentialID{})
 	if err != nil {
-		return cld, cred, errors.Trace(err)
+		return cloud.Cloud{}, cloud.Credential{}, errors.Trace(err)
 	}
-	stmtCACerts, err := p.Prepare(`
-SELECT &CloudCACert.*
-FROM cloud_ca_cert
-WHERE cloud_uuid = $M.uuid`, sqlair.M{}, cloudstate.CloudCACert{})
-	if err != nil {
-		return cld, cred, errors.Trace(err)
+
+	args := sqlair.M{
+		"model_id": modelID,
 	}
+	ids := modelCloudAndCredentialID{}
 
 	var (
-		cloudRows []CloudRow
-		credRows  []credentialstate.Credential
-		keyValues []credentialstate.CredentialAttribute
-	)
-	err = tx.Query(ctx, stmt, sqlair.M{"uuid": modelUUID}).GetAll(
-		&cloudRows,
-		&credRows, &keyValues,
+		cld        cloud.Cloud
+		credResult credential.CloudCredentialResult
 	)
-	if errors.Is(err, sql.ErrNoRows) || len(cloudRows) == 0 || len(credRows) == 0 {
-		return cld, cred, fmt.Errorf("cloud credential for model %q%w", modelUUID, errors.Hide(errors.NotFound))
-	}
-	if err != nil {
-		return cld, cred, fmt.Errorf("querying cloud credential for model %q: %w", modelUUID, err)
-	}
-	cloudRow := cloudRows[0]
-	credRow := credRows[0]
-
-	var dbCloudCACerts []cloudstate.CloudCACert
-	err = tx.Query(ctx, stmtCACerts, sqlair.M{"uuid": cloudRow.ID}).GetAll(&dbCloudCACerts)
-	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
-		return cld, cred, errors.Trace(err)
-	}
-
-	authType := cloud.AuthType(cloudRow.AuthType)
-	isControllerCloud := cloudRow.ModelName == coremodel.ControllerModelName &&
-		cloudRow.ModelOwnerUserName == coremodel.ControllerModelOwnerUsername
-	cld = cloud.Cloud{
-		Name:              cloudRow.Name,
-		Type:              cloudRow.CloudType,
-		AuthTypes:         []cloud.AuthType{authType},
-		Endpoint:          cloudRow.Endpoint,
-		SkipTLSVerify:     cloudRow.SkipTLSVerify,
-		IsControllerCloud: isControllerCloud,
-	}
-	for _, c := range dbCloudCACerts {
-		cld.CACertificates = append(cld.CACertificates, c.CACert)
-	}
-	credAttrs := make(map[string]string, len(keyValues))
-	for _, kv := range keyValues {
-		if kv.Key == "" {
-			continue
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, args).Get(&ids)
+		if errors.Is(err, sql.ErrNoRows) {
+			return fmt.Errorf("%w for id %q", modelerrors.NotFound, modelID)
+		} else if err != nil {
+			return err
+		}
+
+		cld, err = cloudstate.GetCloudForID(ctx, s, tx, ids.CloudID)
+		if err != nil {
+			return fmt.Errorf("getting model %q cloud for id %q: %w", modelID, ids.CloudID, err)
+		}
+
+		credResult, err = credentialstate.GetCloudCredential(ctx, s, tx, ids.CredentialID)
```

Comment:
It was done like this originally and Joe asked for it to be changed to use bespoke SQL and avoid making the api call to the other state package. ie only cherry pick the actual attributes needed for the use case

Reply:
That is fine.... but:

- The logic is very complex and error prone. Hence moving it as this func had several logic bugs in it. When running the tests after fixing it they all failed because they were asserting the bad behaviour.